File: 프로그래머스/5/49190. 방의 개수/방의 개수.java
import java.util.*;

class Solution {
    public int solution(int[] arrows) {
        int answer = 0;
        
        int[] dx = {0, 1, 1, 1, 0,-1,-1,-1};
        int[] dy = {1, 1, 0,-1,-1,-1, 0, 1};
        
        int x = 0;
        int y = 0;
        
        Set<String> nodesVisited = new HashSet<>();
        Set<String> edgesVisited = new HashSet<>();
        
        nodesVisited.add(x + "," + y);
        
        for(int arrow : arrows){
            for(int i = 0; i < 2; i++){
                int nx = x + dx[arrow];
                int ny = y + dy[arrow];
                
                String nodeKey = nx + "," + ny;
                String edgeKey1 = x + "," + y + "," + nx + "," + ny;
                String edgeKey2 = nx + "," + ny + "," + x + "," + y;
                
                if(!edgesVisited.contains(edgeKey1)){
                    if(nodesVisited.contains(nodeKey)){
                        answer++;
                    }
                    edgesVisited.add(edgeKey1);
                    edgesVisited.add(edgeKey2);
                    nodesVisited.add(nodeKey);
                }
                
                x = nx;
                y = ny;
            }
        }
        return answer;
    }
}

File: 프로그래머스/5/49190. 방의 개수/방의 개수.swift
struct Point: Hashable {
    let x: Int
    let y: Int
}

struct Edge: Hashable {
    let from: Point
    let to: Point
}

func solution(_ arrows: [Int]) -> Int {
    let dx = [0, 1, 1, 1, 0, -1, -1, -1]
    let dy = [1, 1, 0, -1, -1, -1, 0, 1]
    var x = 0
    var y = 0
    var answer = 0
    var visitedNodes = Set<Point>()
    var visitedEdges = Set<Edge>()
    let start = Point(x: x, y: y)
    visitedNodes.insert(start)
    for arrow in arrows {
        for _ in 0..<2 {
            let nx = x + dx[arrow]
            let ny = y + dy[arrow]
            let currentNode = Point(x: x, y: y)
            let nextNode = Point(x: nx, y: ny)
            let edge = Edge(from: currentNode, to: nextNode)
            let reverseEdge = Edge(from: nextNode, to: currentNode)
            if !visitedEdges.contains(edge) {
                if visitedNodes.contains(nextNode) {
                    answer += 1
                }
                visitedEdges.insert(edge)
                visitedEdges.insert(reverseEdge)
            }
            visitedNodes.insert(nextNode)
            x = nx
            y = ny
        }
    }
    return answer
}
